src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,887 @@
+--------------------------------------------------------------------------------
+-- | This module is a preliminary part of the implementation of "Proof by 
+--   Logical Evaluation" where we unfold function definitions if they *must* be 
+--   unfolded, to strengthen the environments with function-definition-equalities. 
+--   In this module, we use an interpreter in Haskell to attempt to evaluate down
+--   expressions and generate equalities.
+--------------------------------------------------------------------------------
+
+{-# LANGUAGE DeriveGeneric             #-}
+{-# LANGUAGE OverloadedStrings         #-}
+{-# LANGUAGE PartialTypeSignatures     #-}
+{-# LANGUAGE TupleSections             #-}
+{-# LANGUAGE BangPatterns              #-}
+{-# LANGUAGE FlexibleInstances         #-}
+{-# LANGUAGE ViewPatterns              #-}
+{-# LANGUAGE PatternGuards             #-}
+{-# LANGUAGE RecordWildCards           #-}
+{-# LANGUAGE ExistentialQuantification #-}
+
+module Language.Fixpoint.Solver.Interpreter (instInterpreter) where
+
+import           Language.Fixpoint.Types hiding (simplify)
+import           Language.Fixpoint.Types.Config  as FC
+import qualified Language.Fixpoint.Types.Visitor as Vis
+import qualified Language.Fixpoint.Misc          as Misc 
+import           Language.Fixpoint.Smt.Theories
+import           Language.Fixpoint.Defunctionalize
+import qualified Language.Fixpoint.Utils.Trie    as T 
+import           Language.Fixpoint.Utils.Progress 
+import           Language.Fixpoint.SortCheck
+import           Language.Fixpoint.Graph.Deps             (isTarget) 
+import           Language.Fixpoint.Solver.Sanitize        (symbolEnv)
+import           Language.Fixpoint.Solver.Rewrite
+import           Control.Monad.State
+import           Data.Hashable
+import qualified Data.HashMap.Strict  as M
+import qualified Data.HashSet         as S
+import qualified Data.List            as L
+import qualified Data.Maybe           as Mb
+--import           Debug.Trace                              (trace)
+
+mytracepp :: (PPrint a) => String -> a -> a
+mytracepp = notracepp 
+
+mytrace :: String -> a -> a
+mytrace = {-trace-} flip const
+
+--------------------------------------------------------------------------------
+-- | Strengthen Constraint Environments via PLE 
+--------------------------------------------------------------------------------
+instInterpreter :: (Loc a) => Config -> SInfo a -> Maybe [SubcId] -> IO (SInfo a)
```

Comment:
Either here or in the module documentation, it would help offering context on how this PLE version differs from others.

Just as an example of puzzlement, perhaps this implementation tries to discover equations without any help at all from the SMT solver. This could be made a tad more explicit.

Reply:
Yes, one big part of the interpreter idea was to avoid any interaction with the SMT solver, which also avoids the expensive I/O associated with the communication. I'll add to the documentation